hw1/bashcommands.py

```diff
@@ -0,0 +1,141 @@
+from collections import deque
+from typing import Callable, TypeVar, Optional, Tuple, Union
+from functools import wraps
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+ReceivedFunction = Callable[[T, int, int, R], R]
+ResultFunction = Callable[..., R]
+
+
+def for_each_argument(initial_value: R = None) -> Callable[[ReceivedFunction], ResultFunction]:
+    """
+    This function must be called for a decorator to be returned!
+    This returns a decorator which accepts a function that describes its behaviour for 1 target parameter of type T,
+    the decorator returns a function that accepts any number of target parameters, which iterates over its arguments
+    and for each calls the original function.
+    The returned function is called the result function.
+    :param initial_value: the initial value of the accumulator
+    :return: the decorator that accepts the original function which receives 4 parameters: one of many
+    target arguments of type T, current target argument index in the list of all target arguments provided
+    to the result function, total number of target arguments provided to the result function,
+    accumulator of type R which is equal to the return value of the f function on the previous iteration
+    the result function which accepts any number of target arguments, which returns the value returned by f on
+    the last iteration
+    """
+
+    # this trick is used to remain the return types clear
+    def decorator(f: ReceivedFunction) -> ResultFunction:
+        """
+
+        :param f: The original function which receives 4 parameters: one of many target arguments of type T,
+        current target argument index in the list of all target arguments provided to the result function,
+        total number of target arguments provided to the result function,
+        accumulator of type R which is equal to the return value of the f function on the previous iteration
+        :return:
+        """
+        accumulator = initial_value
+
+        @wraps(f)
+        def f_for_many(*many_args):
+            nonlocal accumulator
+            for arg_index, arg in enumerate(many_args):
+                accumulator = f(arg, arg_index, len(many_args), accumulator)
```

Comment:
`len(many_args)` can be calculated outside of the `for` loop one time.

Reply:
Isn't that a micro optimization? Why do we have to worry about such a thing?

Reply:
It is, it's just something to keep in mind. Some people really care about such things.